### PR TITLE
Allow use of cloud stores for intermediate zarr files

### LINKED ISSUE #7: Allow use of cloud stores for intermediate zarr files
It should be possible to specify a cloud store for the `tempdir` parameter to `vcf_to_zarr`.

### PR DESCRIPTION
**Pull request by tomwhite**
Fixes #7

### CODE REVIEW

File: sgkit_vcf/tests/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+def test_temporary_directory():
+
+    prefix = "prefix-"
+    suffix = "-suffix"
+    with temporary_directory(suffix=suffix, prefix=prefix) as tmpdir:
+
+        dir = Path(tmpdir)
+        assert dir.exists()
+        assert Path(tmpdir).name.startswith(prefix)
```
Comment (jeromekelleher) [inline sgkit_vcf/tests/test_utils.py]: Duplicating ``Path(tmpdir)`` here

File: sgkit_vcf/utils.py [hunk 2]
```diff
@@ -95,3 +100,57 @@ def open_gzip(path: PathType, storage_options: Optional[Dict[str, str]]) -> IO[A
...
+    dir = dir or tempfile.gettempdir()
+    storage_options = storage_options or {}
+
+    # Find the filesystem by looking at the URL scheme (protocol), empty means local filesystem
+    protocol = urlparse(str(dir)).scheme
+    fs = fsspec.filesystem(protocol, **storage_options)
+
+    # Construct a random directory name (use Path purely for filename manipulation, not for IO)
+    tempdir = str(Path(dir) / (prefix + _random_name() + suffix))
+    fs.mkdir(tempdir)
```
Comment (jeromekelleher) [inline sgkit_vcf/utils.py]: Shouldn't we have a rejection sampling loop here to deal with (unlikely, admittedly) collisions?

File: sgkit_vcf/utils.py [hunk 3]
```diff
@@ -95,3 +100,57 @@ def open_gzip(path: PathType, storage_options: Optional[Dict[str, str]]) -> IO[A
...
+    prefix = prefix or ""
+    dir = dir or tempfile.gettempdir()
+    storage_options = storage_options or {}
+
+    # Find the filesystem by looking at the URL scheme (protocol), empty means local filesystem
+    protocol = urlparse(str(dir)).scheme
+    fs = fsspec.filesystem(protocol, **storage_options)
+
+    # Construct a random directory name (use Path purely for filename manipulation, not for IO)
+    tempdir = str(Path(dir) / (prefix + _random_name() + suffix))
```
Comment (jeromekelleher) [inline sgkit_vcf/utils.py]: It's not obvious that ``Path(dir)`` will respond correctly to arbitrary URLs here, like ``s3://`` or whatever. We're not really testing for this either.

**Review by jeromekelleher — COMMENTED** [review]
Looks good @tomwhite, but we're pretty light on testing here. There's a lot of corner cases that can go wrong. I think we need tests for 

- Make sure that the ``dir=file://`` works as expected
- Verify sensible behaviour when ``dir`` is a local file but doesn't exist
- Check that well-formed URLs that are not local files get processes as expected (I'm guessing a mock is the right way to do this: we mock out fspec.filesystem() and check that it's (a) getting called with the right args, and (b) fs.mkdir is getting called with the right args later.
- Verify that name collisions are handled correctly. This should be possible to do by calling tempdir twice with random.seed() set to the same value immediately before calling.

**Comment by tomwhite** [comment]
@jeromekelleher thanks for enumerating those cases. I think most of them are straightforward.

Name collisions are tricky for cloud object stores, since they don't have a first-class concept of directory. Indeed, the doc for [fsspec mkdir](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.spec.AbstractFileSystem.mkdir) states:

> For systems that don’t have true directories, may create an [entry?] for this instance only and not touch the real filesystem

So we could easily get into race conditions and eventual consistency issues.

Instead, I think we should use a [UUID](https://docs.python.org/3/library/uuid.html) (rather than the 8 character random string I used in the first PR) to ensure uniqueness.

File: sgkit_vcf/tests/test_utils.py (hunk 1 above)
Comment (tomwhite) [reply sgkit_vcf/tests/test_utils.py]: Fixed in https://github.com/pystatgen/sgkit-vcf/pull/9/commits/d7ea0012c58ab58410e692b24e89778c769395dd

File: sgkit_vcf/utils.py [hunk 4]
```diff
@@ -95,3 +100,52 @@ def open_gzip(path: PathType, storage_options: Optional[Dict[str, str]]) -> IO[A
...
+    dir = dir or tempfile.gettempdir()
+    storage_options = storage_options or {}
+
+    # Find the filesystem by looking at the URL scheme (protocol), empty means local filesystem
+    protocol = urlparse(str(dir)).scheme
+    fs = fsspec.filesystem(protocol, **storage_options)
+
+    # Construct a random directory name (use Path purely for filename manipulation, not for IO)
+    # The UUID is based on the host ID, current time,and a random component
+    tempdir = str(Path(dir) / (prefix + str(uuid.uuid1()) + suffix))
```
Comment (jeromekelleher) [inline sgkit_vcf/utils.py]: I wonder if we're leaking information here unnecessarily (who knows how the hostname/system time could be used), as if it would be better to use uuid4 altogether. Having the hostname in there makes it no more unique in practice (possibly less, actualy), and the proba of a uuid4 collision is officially negligible.

**Review by jeromekelleher — APPROVED** [review]
LGTM, thanks @tomwhite. UUID is much better - although I wonder if it's better to go all the way with uuid4.

**Comment by tomwhite** [comment]
It turns out that this needed a bit more work to get it to handle non-local filesystems properly, after I tried using it on GCS.
* Pass `storage_options` through API calls.
* Fix (and test) tempdir URL construction (using pytest-mock, and callee).

As a result of these changes I discovered that `file:` URLs didn't work, and on closer inspection it looks like fsspec doesn't support them. More accurately, fsspec *will* get a local filesystem for a `file:` URL, but it *won't* open/mkdir a file/dir for a for a `file:` URL. I propose that we don't support `file:` URLs - at least for the moment.

**Comment by tomwhite** [comment]
(BTW the reason that `file:` URLs were working before was because I was converting them to strings using `str(Path(url))`. But this isn't safe for non-local URLs, e.g. `s3://` or `gs://`. As a rule, I think we should do as little path manipulation as possible.)

**Comment by hammer** [comment]
I'm out of my lane here but is there a reason to manipulate URLs as strings rather than use [urllib.parse](https://docs.python.org/3/library/urllib.parse.html)?

**Comment by hammer** [comment]
> I discovered that file: URLs didn't work, and on closer inspection it looks like fsspec doesn't support them.

Should we file an issue upstream?

**Comment by tomwhite** [comment]
> I'm out of my lane here but is there a reason to manipulate URLs as strings rather than use urllib.parse?

Good question. Actually, urllib.parse is being used to extract the protocol from the URL string. However, [urllib.parse.urljoin](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urljoin) doesn't work for arbitrary protocols (like `s3`).

I wondered if we should use a URI library like [uritools](https://uritools.readthedocs.io/en/stable/#). But even that would require additional utility functions to extract a filename, and combine a parent directory with a child directory or file.

> I discovered that file: URLs didn't work

This turned out to be a bug in the test, and `file:///` URLs do work, so I've added a test for that case.

File: sgkit_vcf/utils.py [hunk 5]
```diff
@@ -95,3 +101,67 @@ def open_gzip(path: PathType, storage_options: Optional[Dict[str, str]]) -> IO[A
...
+    return res
+
+
+@contextmanager
+def temporary_directory(
+    suffix: Optional[str] = None,
+    prefix: Optional[str] = None,
+    dir: Optional[PathType] = None,
+    storage_options: Optional[Dict[str, str]] = None,
+) -> Generator[str, None, None]:
```
Comment (ravwojdyla) [inline sgkit_vcf/utils.py]: Instead of `Generator[str, None, None]` you can say `Iterator[str]`.

File: sgkit_vcf/vcf_reader.py [hunk 6]
```diff
@@ -1,23 +1,23 @@
...
 from sgkit.typing import PathType
-from sgkit_vcf.utils import chunks
+from sgkit_vcf.utils import build_url, chunks, temporary_directory, url_filename
 
 DEFAULT_ALT_NUMBER = 3  # see vcf_read.py in scikit_allel
 
 
 @contextmanager
-def open_vcf(path: PathType) -> VCF:
+def open_vcf(path: PathType) -> Generator[VCF, None, None]:
```
Comment (ravwojdyla) [inline sgkit_vcf/vcf_reader.py]: No need to use `Generator`, see above

File: sgkit_vcf/tests/test_utils.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    fsspec.filesystem.assert_called_once_with("myfs")
+    myfs.mkdir.assert_called_once_with(StartsWith("myfs://path/file/mytmp"))
+    myfs.rm.assert_called_once_with(
+        StartsWith("myfs://path/file/mytmp"), recursive=True
+    )
+
+
+def test_build_url():
+    assert build_url("http://host/path", "subpath") == "http://host/path/subpath"
+    assert build_url("http://host/path/", "subpath") == "http://host/path/subpath"
```
Comment (jeromekelleher) [inline sgkit_vcf/tests/test_utils.py:76]: Be good to test this with more awkard examples:

- Non http scheme (s3, say)
- No scheme, just a local file path like "relative_path/stuff" and "/absolute_path/stuff/"
- A URL with a query string  ``http://host/path?x=1`` -> ``http://host/path/subpath?x=`` presumably?
- A URL with some url encoded entities ``http://host/a%20path/``, say
- A URL that has some things in it that need to be URL encoded, say,  "http://host/!#_%" - presumably, we should always return a properly URL encoded string?

**Comment by tomwhite** [comment]
Hopefully final update - use [yarl](https://yarl.readthedocs.io/en/latest/index.html) for URL parsing (simple API, more popular library). This is ready for review again now.

**Review by jeromekelleher — COMMENTED** [review]
LGTM @tomwhite, I think we just need to do a bit more testing on build_url. There's a lot of different corner cases here, so we need to be reasonably thorough here I think.

File: sgkit_vcf/tests/test_utils.py (hunk 7 above)
Comment (tomwhite) [reply sgkit_vcf/tests/test_utils.py:76]: Thanks @jeromekelleher! Fixed in https://github.com/pystatgen/sgkit-vcf/pull/9/commits/e2de4454a5c465e021f176401a4141a6c303aabc

File: sgkit_vcf/utils.py [hunk 8]
```diff
@@ -95,3 +100,64 @@ def open_gzip(path: PathType, storage_options: Optional[Dict[str, str]]) -> IO[A
...
+    fs = fsspec.filesystem(protocol, **storage_options)
+
+    # Construct a random directory name
+    tempdir = build_url(dir, prefix + str(uuid.uuid4()) + suffix)
+    fs.mkdir(tempdir)
+    try:
+        yield tempdir
+    finally:
+        # Remove the temporary directory on exiting the context manager
+        fs.rm(tempdir, recursive=True)
```
Comment (ravwojdyla) [inline sgkit_vcf/utils.py:163]: I wonder if we should allow to control whether temp files are deleted via a flag (default = True)?

File: sgkit_vcf/utils.py [hunk 9]
```diff
@@ -95,3 +100,64 @@ def open_gzip(path: PathType, storage_options: Optional[Dict[str, str]]) -> IO[A
...
+    # the division (/) operator discards query and fragment, so add them back
+    return str((url / child_path).with_query(url.query).with_fragment(url.fragment))
+
+
+@contextmanager
+def temporary_directory(
+    suffix: Optional[str] = None,
+    prefix: Optional[str] = None,
+    dir: Optional[PathType] = None,
+    storage_options: Optional[Dict[str, str]] = None,
```
Comment (ravwojdyla) [inline sgkit_vcf/utils.py:123]: Are these storage options tested anywhere?

File: sgkit_vcf/utils.py [hunk 10]
```diff
@@ -95,3 +100,64 @@ def open_gzip(path: PathType, storage_options: Optional[Dict[str, str]]) -> IO[A
...
+    dir = dir or tempfile.gettempdir()
+    storage_options = storage_options or {}
+
+    # Find the filesystem by looking at the URL scheme (protocol), empty means local filesystem
+    protocol = urlparse(str(dir)).scheme
+    fs = fsspec.filesystem(protocol, **storage_options)
+
+    # Construct a random directory name
+    tempdir = build_url(dir, prefix + str(uuid.uuid4()) + suffix)
+    fs.mkdir(tempdir)
```
Comment (ravwojdyla) [inline sgkit_vcf/utils.py]: I wonder if this `fs.mkdir` call should be inside the `try/finally`?

**Review by jeromekelleher — APPROVED** [review]
LGTM, reapproving.

**Review by ravwojdyla — APPROVED** [review]

File: sgkit_vcf/utils.py (hunk 5 above)
Comment (tomwhite) [reply sgkit_vcf/utils.py]: Thanks. Fixed in https://github.com/pystatgen/sgkit-vcf/pull/9/commits/56716520f7221b2357d28ba848fdbedef8e7d0c8

File: sgkit_vcf/utils.py (hunk 10 above)
Comment (tomwhite) [reply sgkit_vcf/utils.py]: https://github.com/pystatgen/sgkit-vcf/pull/9/commits/b43f491a0648c4d996ead327b1b0af81ecb39230

File: sgkit_vcf/utils.py (hunk 9 above)
Comment (tomwhite) [reply sgkit_vcf/utils.py:123]: https://github.com/pystatgen/sgkit-vcf/pull/9/commits/7b873b912c2c41be2c2e27dcfa545658b3499f95

File: sgkit_vcf/utils.py (hunk 8 above)
Comment (tomwhite) [reply sgkit_vcf/utils.py:163]: That could be useful. I've opened https://github.com/pystatgen/sgkit-vcf/issues/10 for it.
